[#159] 게시물: 게시물 화면 뷰 구현

jwonyLee opened the pull request:
## 개요

게시물 화면 구현(TableViewCell)

## 동작 화면

<img src="https://user-images.githubusercontent.com/15073405/140017778-0e813057-2101-474a-b34c-f06cdb432254.gif" width="50%" />

## 작업 사항

- UIView+
  - `addSubviews(_:)`: 한 번에 여러 개의 뷰를 서브 뷰로 추가할 수 있게 구현
- LabelWithButtonRoundCollectionCell
  - `configure(text:)`를 수정하여 `removeButton`을 숨김 처리할 수 있게 변경
- LeftRightButtonView
  - 하단 bottomLine 추가
- RatingView
  - 이미지 에셋 변경
  - 이니셜라이저를 통해 버튼 사이 간격을 입력받을 수 있게 변경
- PostTableCell
  - 게시물을 표시하기 위한 뷰. UITableViewCell을 서브클래싱
- PostTableCell+setup
  - PostTableCell의 setup 메서드를 별도로 분리
  - categoryCollectionView.delegate 구현 → 슬라이드 하면 itemCountLabel 업데이트
- PostCategoryViewDataSouce.swift
  - PostTableCell.categoryCollectionView 에 데이터 소스를 별도로 분리한 클래스
- PostSlideImageViewDataSource
  - Post.TableCell.slideImageCollectionView 에 데이터 소스를 별도로 분리한 클래스

## 예외 사항
- 실제 데이터 바인딩X
- moreButton을 탭했을 때 드롭박스 표시 X → #161 에 반영하겠습니당

## 테스트 코드

```swift
import UIKit

final class ViewController: UIViewController {
    // MARK: - View Properties
    private let tableView: UITableView = {
        let tableView: UITableView = UITableView()
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.backgroundColor = .clear
        tableView.separatorStyle = .none
        tableView.allowsSelection = false
        return tableView
    }()

    // MARK: - Properties
    var coordinator: Coordinator?

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = SwiftGenColors.white.color

        setupNavigationBar()
        setupView()
        setupBinding()
    }

    func setupNavigationBar() {
        setupBaseNavigationBar()

        let logoView: LogoView = LogoView("게시물")
        navigationItem.titleView = logoView
    }

    func setupView() {
        setupTableView()
    }

    func setupBinding() {
    }
}

extension ViewController {
    private func setupTableView() {
        view.addSubview(tableView)

        NSLayoutConstraint.activate([
            tableView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
            tableView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor)
        ])

        tableView.estimatedRowHeight = 44
        tableView.rowHeight = UITableView.automaticDimension
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(PostTableCell.self, forCellReuseIdentifier: PostTableCell.reuseIdentifier)
    }
}

extension ViewController: UITableViewDelegate {
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        10
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell: PostTableCell = tableView.dequeueReusableCell(withIdentifier: PostTableCell.reuseIdentifier, for: indexPath) as? PostTableCell else {
            return PostTableCell()
        }

        // 초기 설정
        cell.isTrash = true
        cell.isBought = true

        // 휴지통
        cell.likeButton.rx.tap
            .bind { [weak self] in
                UIView.performWithoutAnimation {
                    self?.tableView.beginUpdates()
                    cell.isTrash = true
                    cell.isBought = false
                    self?.tableView.endUpdates()
                }
            }.disposed(by: cell.disposeBag)

        // 사고싶다
        cell.commentButton.rx.tap
            .bind { [weak self] in
                UIView.performWithoutAnimation {
                    self?.tableView.beginUpdates()
                    cell.isBought = true
                    cell.isTrash = false
                    self?.tableView.endUpdates()
                }
            }.disposed(by: cell.disposeBag)

        // 그 외 default
        cell.photocardButton.rx.tap
            .bind { [weak self] in
                UIView.performWithoutAnimation {
                    self?.tableView.beginUpdates()
                    cell.isTrash = true
                    cell.isBought = true
                    self?.tableView.endUpdates()
                }
            }.disposed(by: cell.disposeBag)

        return cell
    }
}
```

### Linked Issue

close #159

jwonyLee committed:
[#159] feat: UIView.addSubviews(_:)

jwonyLee committed:
[#159] view: PostTableCell 구현

- LeftRightButtonView에 BottomLine 추가
- RatingView 이니셜라이저로 stackView spacing을 입력받게 변경
- LabelWithButtonRoundCollectionCell.configure 메서드에서 removeButton을 숨김처리할 수 있게 변경
- PostCategoryViewDataSource, PostSlideImageViewDataSource -> PostTableCell에 들어가는 컬렉션 뷰에 데이터 소스 구현
- 게시물을 보여줄 PostTableCell 구현

jwonyLee committed:
[#159] feat: 휴지통, 사고싶다 일 때 버튼 숨김/표시하는 메서드 구현

jwonyLee committed:
[#159] feat: 이미지를 슬라이드할 때 itemCountLabel 업데이트

jwonyLee committed:
[#159] docs: 프로퍼티 주석 추가

jwonyLee committed:
[#159] view: nameLabel, placeLabel 글자가 길어질 때를 대비한 스크롤뷰 구현

gustn3965 commented on ThingLog/View/LeftRightButtonView.swift:
```diff
@@ -83,7 +93,11 @@ class LeftRightButtonView: UIView {
+            bottomLineView.leadingAnchor.constraint(equalTo: leadingAnchor),
+            bottomLineView.trailingAnchor.constraint(equalTo: trailingAnchor),
+            bottomLineView.topAnchor.constraint(equalTo: bottomAnchor)
         ])
```
`bottomLineView`는 `bottomAnchor`에 붙어있고,` stackView`도 `bottomAnchor`에 붙어있어서, ` bottomLineView`와 `stackView`사이에 정확한 지점이 없어서 간혹 뷰가 이렇게 나오는 경우가 있습니다. 
![image](https://user-images.githubusercontent.com/48749182/140041602-115d06f9-7ffc-4066-a77e-622746a6a444.png)

그리고 `LeftRightButtonView`를 `bottomLineView`를 추가하셨기 때문에,` TrashViewController`에서는 다음과 같이 ( 기존대로 ) 화면이 잘 나오도록 해주시면 됩니당. 
![image](https://user-images.githubusercontent.com/48749182/140042349-b9076b8a-2df5-4b07-a839-5cd99fe34a6b.png)

gustn3965 commented on ThingLog/View/TableVeiwCell/PostTableCell/PostTableCell.swift:
```diff
@@ -0,0 +1,316 @@
+    let headerContainerView: UIView = {
+        let view: UIView = UIView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .clear
```
`PostTableCell`은 보시다시피 너무 많은 뷰를 가지고 있는 듯 합니다. 이런 요소는 앞으로 리팩토링이나 뷰 추가 등에 있어서 굉장히 이해하는데 어려울 것 같아요. 그래서 지금 이렇게 많은 뷰 프로퍼티를 한번에 다 나열해서 갖는게 좋은 건지 잘 모르겠습니다. 저는 이 뷰에서만 사용하더라도 최대한 알기 쉽게끔 분리하는게 좋을 것 같다라는 생각입니다만..  **일단 이부분은 당장은 넘어가셔도 괜찮을 것 같습니다.** 

그리고 뷰와 뷰이외 프로퍼티 사이의 `MARK` 주석을 추가하여 구분해주는 것도 좋을 것 같습니다. 

그리고 `PostTableCell`이 유니크하더라도 노션에 이미지 링크를 추가해주시는 것도 좋을 것 같습니다.

gustn3965 commented on ThingLog/View/TableVeiwCell/PostTableCell/PostCategoryViewDataSouce.swift:
```diff
@@ -0,0 +1,28 @@
+        10
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell: LabelWithButtonRoundCollectionCell = collectionView.dequeueReusableCell(withReuseIdentifier: LabelWithButtonRoundCollectionCell.reuseIdentifier, for: indexPath) as? LabelWithButtonRoundCollectionCell else {
+            return LabelWithButtonRoundCollectionCell()
+        }
+
+        cell.configure(text: "전자제품", buttonIsHidden: true)
```
새로운 GUI에서는 카테고리 표시하는 부분이 다음과 같이 바꼈으므로, 변경하셔야 할 것 같습니다. 
![image](https://user-images.githubusercontent.com/48749182/140270115-112d98e9-c2b8-48b9-be39-21e82bec61de.png)

gustn3965 commented on ThingLog/View/TableVeiwCell/PostTableCell/PostTableCell+setup.swift:
```diff
@@ -0,0 +1,196 @@
+        ])
+    }
+
+    private func createCategoryCollectionViewLayout() -> UICollectionViewCompositionalLayout {
+        let itemSize: NSCollectionLayoutSize = .init(widthDimension: .estimated(1), heightDimension: .estimated(1))
+        let item: NSCollectionLayoutItem = .init(layoutSize: itemSize)
+
+        let groupSize: NSCollectionLayoutSize = .init(widthDimension: .estimated(1),
+                                                      heightDimension: .fractionalHeight(1))
+        let group: NSCollectionLayoutGroup = .horizontal(layoutSize: groupSize, subitems: [item])
+
+        let section: NSCollectionLayoutSection = .init(group: group)
+        section.orthogonalScrollingBehavior = .continuous
+        section.interGroupSpacing = 10
+        section.contentInsets = .init(top: 8, leading: 10, bottom: 0, trailing: 10)
+
+        let layout: UICollectionViewCompositionalLayout = .init(section: section)
+
+        return layout
+    }
```
워닝이 많이 뜨는 곳이 카테고리 컬렉션뷰에서 많이 뜨더라구요.  ( 사실 안고쳐도 뷰에는 큰 이상은 없어보입니다. )
그래서 뷰 디버깅하면서 최대한 고쳐봤습니다. 아래와 같이 코드로 고쳐보니 해당 셀에서 발생하는 오토레이아웃 워닝은 사라지는 걸로 볼 수 있었습니다. ( 나머지 오토레이아웃 워닝은, `deleteButton`에서 생기는 것 같습니다. 해당 부분은 코멘트로 언급해드렸습니다. ) 
주된 원인은 `itemSize`의 `width`가 `.estimated(1)`이 원인인 것 같습니다. 
더 정확하게는 .estimated가 16보다 작은 경우에는 워닝이 뜨는 것 같습니다.  이유를 추정해보니, `LabelWithButtonRoundCollectionCell`에서 `StackView`가 좌우로 `8`씩 `constriant`를 주었기 때문에, 최소한 `8 * 2`의 값은 `.estimated`로 가져야 하는 것 같습니다. 

추가적으로 `itemSize`의 `height`는 `fractionHeight`로, 섹션 `inset`도 `bottom`에 `8`을 추가한걸로 변경했습니다. 아무래도 섹션에 탑만 `8`있고, `label`의` intrinsicContentSize`로만 판단하여 셀이 가운데에 나타나게 하는 것은 그다지 직관적이지 않는 것 같아서요.

```swift
    private func createCategoryCollectionViewLayout() -> UICollectionViewCompositionalLayout {
        let itemSize: NSCollectionLayoutSize = .init(widthDimension: .estimated(16),
                                                     heightDimension: .fractionalHeight(1))
        let item: NSCollectionLayoutItem = .init(layoutSize: itemSize)

        let groupSize: NSCollectionLayoutSize = .init(widthDimension: .estimated(1),
                                                      heightDimension: .fractionalHeight(1))
        let group: NSCollectionLayoutGroup = .horizontal(layoutSize: groupSize, subitems: [item])

        let section: NSCollectionLayoutSection = .init(group: group)
        section.orthogonalScrollingBehavior = .continuous
        section.interGroupSpacing = 10
        section.contentInsets = .init(top: 8, leading: 10, bottom: 8, trailing: 10)

        let layout: UICollectionViewCompositionalLayout = .init(section: section)

        return layout
    }
```

gustn3965 reviewed (commented):
굉장히 빠른 시간안에 복잡한 뷰를 잘 구현하신 것 같습니다. 

코멘트외의 추가적으로, 컨텐츠 내용이 없는 경우에 회색과 default 텍스트가 필요합니다. 
![image](https://user-images.githubusercontent.com/48749182/140270243-bf562a15-e3f7-4775-814b-72d7daec1a1e.png)

jwonyLee committed:
[#159] ref: 피드백 반영
- LabelWithButtonRoundCollectionCell borderLine 추가
- RatingView.currentRating 값이 변경될 때 이미지 변경
- (현수님) createCategoryCollectionViewLayout 수정 사항 반영
- PostTableCell 주석 추가

jwonyLee committed:
[#159] ref: HorizontalScrollLabel 구현

jwonyLee committed:
[#159] ref: LeftRightButtonView bottomLine 수정

jwonyLee commented:
다음과 같은 수정사항이 있습니다.
- LeftRightButtonView의 이니셜라이저로 bottomLineView의 표시 여부를 받고, 해당 값에 따라 오토레이아웃을 따로 설정하게 했습니다.  기본값으로 bottomeLineView를 숨기므로 TrashViewController에서 별도 수정 없이 기존처럼 표시됩니다.
- RatingView.currentRating 값이 변경될 때 이미지를 변경하는 코드를 추가했습니다.
- createCategoryCollectionViewLayout 수정 사항 반영했습니다. 감사합니다 XD
- PostTableCell에서 스크롤 뷰로 감싼 레이블을 HorizontalScrollLabel로 구현해서 분리했습니다.
- LabelWithButtonRoundCollectionCell을 새로운 GUI 시안에 맞게 수정했습니다.

추가적으로...
컨텐츠 내용이 없는 경우에 색상 지정, default 텍스트는 실제 데이터 바인딩 구현할 때 반영하려 합니다.

gustn3965 commented on ThingLog/View/TableVeiwCell/PostTableCell/PostTableCell+setup.swift:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    func setupSpecificActionContainerView() {
+        let trashButtonHeight: CGFloat = 54.0
+        let boughtButtonHeight: CGFloat = 52.0
+        let boughtButtonLeadingTrailingSpacing: CGFloat = 20.0
+
+        NSLayoutConstraint.activate([
+            boughtButton.leadingAnchor.constraint(equalTo: specificActionContainerView.leadingAnchor, constant: boughtButtonLeadingTrailingSpacing),
+            boughtButton.topAnchor.constraint(equalTo: specificActionContainerView.topAnchor),
+            boughtButton.trailingAnchor.constraint(equalTo: specificActionContainerView.trailingAnchor, constant: -boughtButtonLeadingTrailingSpacing),
+            boughtButton.bottomAnchor.constraint(equalTo: specificActionContainerView.bottomAnchor),
+            boughtButton.heightAnchor.constraint(equalToConstant: boughtButtonHeight),
+
+            trashActionButton.leadingAnchor.constraint(equalTo: specificActionContainerView.leadingAnchor),
+            trashActionButton.topAnchor.constraint(equalTo: specificActionContainerView.topAnchor),
+            trashActionButton.trailingAnchor.constraint(equalTo: specificActionContainerView.trailingAnchor),
+            trashActionButton.bottomAnchor.constraint(equalTo: specificActionContainerView.bottomAnchor),
+            trashActionButton.heightAnchor.constraint(equalToConstant: trashButtonHeight)
+        ])
+        boughtButton.layer.cornerRadius = boughtButtonHeight / 2
```
지금 `LeftRightButtonView`와, `RoundCenterTextButton` 에서 오토레이아웃 워닝이 뜨고 있습니다. 
로그를 살펴보면, `LeftRightButtonView`의 `HeightConstraint` = 54 를 `Break`하고 있다고 나와있습니다. 그래서, 실제로 뷰디버깅을 통해서 보면, `LeftRightButtonView`의 높이가 54가 아닌, 52로 나오고 있습니다. 

그 이유는, `specificActionContainerView` 라는 `UIView`안에, `LeftRightButton`, `RoundCenterTextButton` 둘다 넣으면서, 둘다 `specificActionContainerView`의 좌우위아래 모두 맞게 Constraint를 지정하면서, 그 둘의 높이를 다르게 주었기 때문에 `specificActionContainerView`가 어떤 높이를 맞추어야할지 모르기 때문입니다.

해결방법은 제 생각에는 두가지 정도 있는 것 같습니다. 하나는 둘다 높이를 54로 맞추거나,  다른 하나는 `specificActionContainerView`를 StackView로 만드는 방법입니다. 

```swift
    /// 특별한 상황(사고싶다, 휴지통)에서 쓰이는 샀어요 버튼, 삭제/복구 버튼을 포함한 뷰
    lazy var specificActionContainerView: UIStackView = {
        let view: UIStackView = UIStackView(arrangedSubviews: [boughtButton, trashActionButton])
        view.axis = .vertical
        view.alignment = .center
        return view
    }()

// ...

    func setupSpecificActionContainerView() {
        let trashButtonHeight: CGFloat = 54.0
        let boughtButtonHeight: CGFloat = 52.0
        let boughtButtonLeadingTrailingSpacing: CGFloat = 20.0

        NSLayoutConstraint.activate([
            boughtButton.leadingAnchor.constraint(equalTo: specificActionContainerView.leadingAnchor, constant: boughtButtonLeadingTrailingSpacing),
            boughtButton.trailingAnchor.constraint(equalTo: specificActionContainerView.trailingAnchor, constant: -boughtButtonLeadingTrailingSpacing),
            boughtButton.heightAnchor.constraint(equalToConstant: boughtButtonHeight),

            trashActionButton.leadingAnchor.constraint(equalTo: specificActionContainerView.leadingAnchor),
            trashActionButton.trailingAnchor.constraint(equalTo: specificActionContainerView.trailingAnchor),
            trashActionButton.heightAnchor.constraint(equalToConstant: trashButtonHeight)
        ])
        boughtButton.layer.cornerRadius = boughtButtonHeight / 2
    }
```

gustn3965 commented on ThingLog/View/TableVeiwCell/PostTableCell/PostTableCell.swift:
```diff
@@ -0,0 +1,300 @@
+        setupStackView()
+    }
+
+    func setupBinding() {
+    }
```
빈 메서드는 따로 작성하지 않는 것이 좋을 것 같습니다.

gustn3965 reviewed (approved):
LGTM,

언급드린 오토레이아웃 워닝은 충분히 해결할 수 있는 부분이라고 생각하여 코멘트 달았습니다.
오토레이아웃 워닝 테스트해보신 다음, 워닝 없애고 Merge하시면 될 것 같습니다.

gustn3965 commented:
아, 추가적으로, View에 대한 이미지 링크 삽입하실 떄, 앞에 `!` 를 없애야 보입니다! 

아래는 `!` 가 있어서 보이지 않습니다
![image](https://user-images.githubusercontent.com/48749182/140694358-4f841762-385b-48b9-be3b-9cf077ea39a1.png)

jwonyLee committed:
[#159] ref: 피드백 반영
- 주석에 느낌표 제거
- 빈 메서드 제거
- boughtButton, trashActionButton 높이 통일

jwonyLee commented:
말씀해주신 내용들 전부 반영했습니다~
`boughtButton`, `trashActionButton` 높이는 병연님이랑 이야기해서 52로 통일 시키기로 했습니다!

jwonyLee committed:
Merge branch 'develop' into view/159-PostCell
